cmd/trace-agent/subcommands/run/service_windows.go

```diff
@@ -0,0 +1,64 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package run
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"golang.org/x/sys/windows/svc"
+
+	"github.com/DataDog/datadog-agent/comp/trace/config"
+	tracecfg "github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type myservice struct {
+	cliParams *RunParams
+	config    config.Component
+}
+
+func (m *myservice) Execute(args []string, r <-chan svc.ChangeRequest, changes chan<- svc.Status) (ssec bool, errno uint32) {
+	const cmdsAccepted = svc.AcceptStop | svc.AcceptShutdown
+	changes <- svc.Status{State: svc.StartPending}
+	changes <- svc.Status{State: svc.Running, Accepts: cmdsAccepted}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	go func() {
+		for {
+			select {
+			case c := <-r:
+				switch c.Cmd {
+				case svc.Interrogate:
+					changes <- c.CurrentStatus
+					// Testing deadlock from https://code.google.com/p/winsvc/issues/detail?id=4
+					time.Sleep(100 * time.Millisecond)
+					changes <- c.CurrentStatus
```

Comment:
So to handle the deadlock, we're sleeping, then adding the same status to the channel again? I'm confused by this

Reply:
This is not my code katie, this is the same pattern used by all windows services:
- https://github.com/DataDog/datadog-agent/blob/main/cmd/dogstatsd/main_windows.go#L115
- https://github.com/DataDog/datadog-agent/blob/main/cmd/process-agent/main_windows.go#L49
- https://github.com/DataDog/datadog-agent/blob/main/cmd/trace-agent/main_windows.go#L47 (even the trace-agent as you see).

And a few others. I am not sure of the details, but this has been around for a while.